bootstrap/src/main/java/jenkins/bootstrap/DependenciesTxt.java

```diff
@@ -0,0 +1,81 @@
+package jenkins.bootstrap;
+
+import hudson.util.VersionNumber;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Parses {@code dependencies.txt} produced by Maven dependency:list-dependencies mojo.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+class DependenciesTxt {
+    final List<Dependency> dependencies;
+    /**
+     * 'groupId:artifactId' to version number
+     */
+    private final Map<String,VersionNumber> jars = new HashMap<>();
+
+    public DependenciesTxt(InputStream i) throws IOException {
+        List<Dependency> list = new ArrayList<>();
+
+        try (Reader r = new InputStreamReader(i,"UTF-8");
+             BufferedReader br = new BufferedReader(r)) {
+
+            // line that we care about has 5 components: groupId:artifactId:packaging[:classifier]:version:scope
+            String line;
+            while ((line=br.readLine())!=null) {
+                String[] tokens = line.trim().split(":");
+
+                Dependency d;
+                switch (tokens.length) {
+                case 5:
+                    d = new Dependency(tokens[0], tokens[1], tokens[3], null);
+                    break;
+                case 6:
+                    d = new Dependency(tokens[0], tokens[1], tokens[4], tokens[3]);
+                    break;
+                default:
+                    continue;
+                }
+                list.add(d);
+                jars.put(d.ga,d.vv);
+            }
+        }
+
+        dependencies = Collections.unmodifiableList(list);
+    }
+
+    /**
+     * Attempts to reverse map a jar file name back to {@link Dependency}
+     */
+    public Dependency fromFileName(String jarFileName) {
```

Comment:
well when I see no annotations I assume that means Nullable. Iow the caller is responsible for knowing when null is appropriate.

If the caller is abdicated of that responsible then CheckForNull is useful to indicate the method is taking the responsibility for the caller. 

NonNull is saying never give me a null.

As the default is Nullable, I do not see adding annotations as being ever a blocking issue with a PR review.

That said, there are times where adding annotations is very useful, and gentle reminders are ok. For example when defining an API / SPI I would see annotations as almost mandatory. But when creating an implementation of the api/spi or when writing code that is not intended to be used elsewhere I see them as very optional.

That's my €0.02 anyway